tests/test_ext_napoleon_docstring.py

```diff
@@ -1971,6 +1977,228 @@ def test_list_in_parameter_description(self):
         actual = str(NumpyDocstring(docstring, config))
         self.assertEqual(expected, actual)
 
+    def test_token_type(self):
+        tokens = (
+            ("1", "literal"),
+            ("'string'", "literal"),
+            ('"another_string"', "literal"),
+            ("{1, 2}", "literal"),
+            ("{'va{ue', 'set'}", "literal"),
+            ("optional", "control"),
+            ("default", "control"),
+            (", ", "delimiter"),
+            (" of ", "delimiter"),
+            (" or ", "delimiter"),
+            (": ", "delimiter"),
+            ("True", "obj"),
+            ("None", "obj"),
+            ("name", "obj"),
+            (":py:class:`Enum`", "reference"),
+        )
+
+        for token, expected in tokens:
+            actual = _token_type(token)
+            self.assertEqual(expected, actual)
+
+    def test_token_type_invalid(self):
+        tokens = (
+            "{1, 2",
+            "}",
+            "'abc",
+            "def'",
+            '"ghi',
+            'jkl"',
+        )
+        for token in tokens:
+            # TODO: check for the warning
+            _token_type(token)
+
+        assert False
+
+    def test_tokenize_type_spec(self):
+        types = (
+            "str",
+            "int or float or None",
+            '{"F", "C", "N"}',
+            "{'F', 'C', 'N'}",
+            '"ma{icious"',
+            r"'with \'quotes\''",
+        )
+        modifiers = (
+            "",
+            "optional",
+            "default: None",
+        )
+
+        type_tokens = (
+            ["str"],
+            ["int", " or ", "float", " or ", "None"],
+            ["{", '"F"', ", ", '"C"', ", ", '"N"', "}"],
+            ["{", "'F'", ", ", "'C'", ", ", "'N'", "}"],
+            ['"ma{icious"'],
+            [r"'with \'quotes\''"],
+        )
+        modifier_tokens = (
+            [],
+            ["optional"],
+            ["default", ": ", "None"],
+        )
+
+        type_specs = tuple(
+            ", ".join([type_, modifier]) if modifier else type_
+            for type_ in types
+            for modifier in modifiers
+        )
+        tokens = tuple(
+            tokens_ + ([", "] + modifier_tokens_ if modifier_tokens_ else [])
+            for tokens_ in type_tokens
+            for modifier_tokens_ in modifier_tokens
+        )
+
+        for type_spec, expected in zip(type_specs, tokens):
```

Comment:
sure. How do you feel about using `pytest.mark.parametrize` instead? However, I'd probably need a non `unittest.TestCase` test in order to be able to use that.

Reply:
I can't answer to move to pytest because the maintainer of napoleon is now absent. Personally, +1 for the move.

Anyway, I think the most important thing is not generating the expected value. Is it really needed to confirm all combination of types and modifiers? How about choosing some examples from the combinations?

Reply:
not sure, I can try reducing that to a few manually generated examples

Reply:
done